src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,333 @@
+        }
+
+        protected override void OnManifestCompose(IsHardforkEnabledDelegate hfChecker, uint blockHeight, ContractManifest manifest)
+        {
+            manifest.SupportedStandards = new[] { "NEP-27" };
+        }
+
+        /// <summary>
+        /// Verify checks whether the transaction is signed by one of the notaries and
+        /// ensures whether deposited amount of GAS is enough to pay the actual sender's fee.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="sig">Signature</param>
+        /// <returns>Whether transaction is valid.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        private bool Verify(ApplicationEngine engine, byte[] sig)
+        {
+            Transaction tx = (Transaction)engine.ScriptContainer;
+            if (tx.GetAttribute<NotaryAssisted>() is null) return false;
+            foreach (var signer in tx.Signers)
+            {
+                if (signer.Account == Hash)
+                {
+                    if (signer.Scopes != WitnessScope.None) return false;
+                    break;
+                }
+            }
+            if (tx.Sender == Hash)
+            {
+                if (tx.Signers.Length != 2) return false;
+                var payer = tx.Signers[1].Account;
+                var balance = GetDepositFor(engine.SnapshotCache, payer);
+                if (balance is null || balance.Amount.CompareTo(tx.NetworkFee + tx.SystemFee) < 0) return false;
+            }
+            ECPoint[] notaries = GetNotaryNodes(engine.SnapshotCache);
+            var hash = tx.GetSignData(engine.GetNetwork());
+            return notaries.Any(n => Crypto.VerifySignature(hash, sig, n));
+        }
+
+        /// <summary>
+        /// OnNEP17Payment is a callback that accepts GAS transfer as Notary deposit.
+        /// It also sets the deposit's lock height after which deposit can be withdrawn.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="from">GAS sender</param>
+        /// <param name="amount">The amount of GAS sent</param>
+        /// <param name="data">Deposit-related data: optional To value (treated as deposit owner if set) and Till height after which deposit can be withdrawn </param>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.All)]
+        private void OnNEP17Payment(ApplicationEngine engine, UInt160 from, BigInteger amount, StackItem data)
+        {
+            if (engine.CallingScriptHash != GAS.Hash) throw new InvalidOperationException(string.Format("only GAS can be accepted for deposit, got {0}", engine.CallingScriptHash.ToString()));
+            var to = from;
+            var additionalParams = (Array)data;
+            if (additionalParams.Count() != 2) throw new FormatException("`data` parameter should be an array of 2 elements");
```

Comment:
`OnNEP17Payment ` may be called by the contract.
So the `data` may not be an `Array`, and this line will throw an `InvalidCastException` if not an `Array`.
It's OK even if `InvalidCastException`?

```suggestion
            var additionalParams = data as Array;
            if (additionalParams  is null || additionalParams.Count() != 2)
                throw new FormatException("`data` parameter should be an array of 2 elements");

```

Reply:
This should be `object` type aka `Any`

Reply:
> OnNEP17Payment may be called by the contract.

It's impossible because several lines below we have a check that ensures that calling context is GasToken contract.

> So the data may not be an Array

But even when the caller is GasToken, it's possible that data has an unexpected format, so you're right here.

> It's OK even if InvalidCastException?

Yes, it's OK, the main point here is that exception will be thrown anyway and execution is aborted. But from user's PoW it's always nice to see a well-formatted exception message instead of just InvalidCastException, hence I've updated this code.